packages/react-i18n/src/index.tsx

```diff
@@ -88,15 +82,27 @@ type Optionalize< T extends K, K > = Omit< T, keyof K >;
  * export default withI18n( MyComponent );
  */
 export const withI18n = < T extends I18nProps = I18nProps >(
-	WrappedComponent: React.ComponentType< T >
-): React.FunctionComponent< Optionalize< T, I18nProps > > => props => {
-	const i18n = useI18n();
-	return (
-		<WrappedComponent
-			{ ...i18n }
-			// Required cast `props as T`
-			// See https://github.com/Microsoft/TypeScript/issues/28938
-			{ ...( props as T ) }
-		/>
+	InnerComponent: React.ComponentType< T >
+): React.FunctionComponent< Optionalize< T, I18nProps > > => {
+	const WrappedComponent: React.FunctionComponent< Optionalize< T, I18nProps > > = props => (
+		<I18nContext.Consumer>
+			{ i18n => {
+				if ( i18n === undefined ) {
+					throw new Error( 'withI18n must be descendent of <I18nProvider />' );
+				}
+				return (
+					<InnerComponent
+						{ ...i18n }
```

Comment:
Why did you decide to abandon the `useI18n` hook here? It's a common practice these days to implement HOC wrappers using hooks.

Reply:
This was mostly related to the other point. I noticed `withI18n` wasn't working well with a `PureComponent` and tried a few different things. I've gone back to the `useI18n` form.